src/com/amazon/corretto/crypto/provider/AesGcmSpi.java

```diff
@@ -675,31 +680,21 @@ private byte[] engineDecryptFinal(byte[] input, final int offset, final int leng
         input = EMPTY_ARRAY;
       }
 
-      // If we already have processed some amount of ciphertext then we must coallesce it prior to
-      // decryption.
-      // Otherwise, we can just use the provided input.
+      checkOutputBuffer(inputLen, output, outputOffset, true);
+      checkArrayLimits(input, inputOffset, inputLen);
+
       final byte[] workingInputArray;
       final int workingInputOffset;
       final int workingInputLength;
-      if (decryptInputBuf.size() > 0) {
-        // We've already processed data so need to merge this in
-        engineUpdate(input, offset, length); // Decrypt mode never generates output for updates
-        workingInputArray = decryptInputBuf.getDataBuffer();
-        workingInputLength = decryptInputBuf.size();
-        workingInputOffset = 0;
-      } else {
-        workingInputArray = input;
-        workingInputLength = length;
-        workingInputOffset = offset;
-      }
+      engineUpdate(input, inputOffset, inputLen);
+      workingInputArray = decryptInputBuf.getDataBuffer();
+      workingInputLength = decryptInputBuf.size();
+      workingInputOffset = 0;
 
       if (workingInputLength < tagLength) {
         throw new AEADBadTagException("Input too short - need tag");
       }
 
-      final int expectedOutputLength = workingInputLength - tagLength;
-      final byte[] result = new byte[expectedOutputLength];
```

Comment:
This is where the memory gets saved. Previously, when the caller provides the buffer this intermediate buffer gets allocated no matter what.